test/test.oauth2common.ts

```diff
@@ -0,0 +1,415 @@
+        url: 'https://www.example.com/path/to/api',
+        method: 'POST',
+        headers: {
+          'Content-Type': 'application/json',
+        },
+        data: {
+          key1: 'value1',
+          key2: 'value2',
+        },
+      };
+      const actualOptions = Object.assign({}, originalOptions);
+      const expectedOptions = Object.assign({}, originalOptions);
+      expectedOptions.data.client_id = reqBodyAuth.clientId;
+      expectedOptions.data.client_secret = reqBodyAuth.clientSecret;
+
+      handler.testApplyClientAuthenticationOptions(actualOptions);
+      assert.deepStrictEqual(expectedOptions, actualOptions);
+    });
+
+    it('should inject secretless creds in json content', () => {
+      const handler = new TestOAuthClientAuthHandler(reqBodyAuthNoSecret);
+      const originalOptions: GaxiosOptions = {
+        url: 'https://www.example.com/path/to/api',
+        method: 'POST',
+        headers: {
+          'Content-Type': 'application/json',
+        },
+        data: {
+          key1: 'value1',
+          key2: 'value2',
+        },
+      };
+      const actualOptions = Object.assign({}, originalOptions);
+      const expectedOptions = Object.assign({}, originalOptions);
+      expectedOptions.data.client_id = reqBodyAuthNoSecret.clientId;
+      expectedOptions.data.client_secret = '';
+
+      handler.testApplyClientAuthenticationOptions(actualOptions);
+      assert.deepStrictEqual(expectedOptions, actualOptions);
+    });
+
+    it('should inject creds in empty json content', () => {
+      const handler = new TestOAuthClientAuthHandler(reqBodyAuth);
+      const originalOptions: GaxiosOptions = {
+        url: 'https://www.example.com/path/to/api',
+        method: 'POST',
+        headers: {
+          'Content-Type': 'application/json',
+        },
+      };
+      const actualOptions = Object.assign({}, originalOptions);
+      const expectedOptions = Object.assign({}, originalOptions);
+      expectedOptions.data = {
+        client_id: reqBodyAuth.clientId,
+        client_secret: reqBodyAuth.clientSecret,
+      };
+
+      handler.testApplyClientAuthenticationOptions(actualOptions);
+      assert.deepStrictEqual(expectedOptions, actualOptions);
+    });
+
+    it('should inject creds in non-empty x-www-form-urlencoded content', () => {
+      const handler = new TestOAuthClientAuthHandler(reqBodyAuth);
+      const originalOptions: GaxiosOptions = {
+        url: 'https://www.example.com/path/to/api',
+        method: 'POST',
+        headers: {
+          // Handling of headers should be case insensitive.
+          'content-Type': 'application/x-www-form-urlencoded',
+        },
+        data: 'key1=value1&key2=value2',
+      };
+      const actualOptions = Object.assign({}, originalOptions);
+      const expectedOptions = Object.assign({}, originalOptions);
+      const credentials =
+        `client_id=${encodeURIComponent(reqBodyAuth.clientId)}` +
```

Comment:
likewise, might be worth considering using querystring if it makes the code look cleaner, I'm open to template strings if it doesn't seem to look much cleaner.

Reply:
Done here and elsewhere.